scipy/special/_lambertw.h

```diff
@@ -0,0 +1,167 @@
+/* Implementation of the Lambert W function [1]. Based on MPMath
+ *  Implementation [2], and documentation [3].
+ *
+ * Copyright: Yosef Meller, 2009
+ * Authoer email: mellerf@netvision.net.il
+ *
+ * Distributed under the same license as SciPy
+ * Translated into C++ by SciPy developers, 2023.
+ *
+ * References:
+ * [1] On the Lambert W function, Adv. Comp. Math. 5 (1996) 329-359,
+ *     available online: http://www.apmaths.uwo.ca/~djeffrey/Offprints/W-adv-cm.pdf
+ * [2] mpmath source code, Subversion revision 990
+ *     http://code.google.com/p/mpmath/source/browse/trunk/mpmath/functions.py?spec=svn994&r=992
+ * [3] mpmath source code, Subversion revision 994
+ *     http://code.google.com/p/mpmath/source/browse/trunk/mpmath/function_docs.py?spec=svn994&r=994
+
+ * TODO: use a series expansion when extremely close to the branch point
+ * at `-1/e` and make sure that the proper branch is chosen there.
+ */
+
+
+#ifndef LAMBERTW_H
+#define LAMBERTW_H
+
+#include <cmath>
+#include <complex>
+#include <numpy/npy_math.h>
+
+#include "_evalpoly.h"
+#include "sf_error.h"
```

Comment:
The `#include "sf_error.h"` has similar issues as to the NumPy header. Additionally, it's not even a header, as there is a `sf_error.c`. Even if it was a header, it can't work in other array libraries as is, as I think I saw it making calls to the Python C API.

I'm not saying we need to fix this now, probably it should be done in a follow-up PR, but there needs to be thought into how error handling is going to work in general.

Reply:
Ah, right. The best thing I can think of is to put checks in the wrapper that has the numpy stuff.

Reply:
Happy to look into this properly with you, but I guess we should probably do it in PR #2.

Reply:
yeah, agreed on punting this for later.